calico_upgrade/Makefile

```diff
@@ -0,0 +1,339 @@
+	# Map the docker socket in so docker can be used from inside the container
+	# All of code under test is mounted into the container.
+	#   - This also provides access to calico-upgrade and the docker client
+	docker run --net=host --privileged \
+	           -e MY_IP=$(LOCAL_IP_ENV) \
+	           --rm -ti \
+	           -v $(SOURCE_DIR):/code \
+	           calico/test \
+	           sh -c 'nosetests $(ST_TO_RUN) -sv --nologcapture  --with-xunit --xunit-file="/code/nosetests.xml" --with-timer $(ST_OPTIONS)'
+
+	$(MAKE) stop-etcd
+
+## Run the STs in a container
+.PHONY: st
+testenv: dist/calico-upgrade dist/calicoctl dist/calicoctlv2 run-etcd
+	# Use the host, PID and network namespaces from the host.
+	# Privileged is needed since 'calico node' write to /proc (to enable ip_forwarding)
+	# Map the docker socket in so docker can be used from inside the container
+	# All of code under test is mounted into the container.
+	#   - This also provides access to calico-upgrade and the docker client
+	docker run --net=host --privileged \
+	           -e MY_IP=$(LOCAL_IP_ENV) \
+	           --rm -ti \
+	           -v $(SOURCE_DIR):/code \
+	           --name=testenv \
+	           calico/test \
+	           sh
+	docker exec -it --name=testenv sh
+
+dist/calicoctl:
+	-docker rm -f calicoctl
+	docker pull $(CTL_CONTAINER_NAME)
+	docker create --name calicoctl $(CTL_CONTAINER_NAME)
+	docker cp calicoctl:calicoctl dist/calicoctl && \
+	  test -e dist/calicoctl && \
+	  touch dist/calicoctl
+	-docker rm -f calicoctl
+
+dist/calicoctlv2:
+	-docker rm -f calicoctlv2
+	docker pull $(CTL_CONTAINER_V2_NAME)
+	docker create --name calicoctlv2 $(CTL_CONTAINER_V2_NAME)
+	docker cp calicoctlv2:calicoctl dist/calicoctlv2 && \
+	  test -e dist/calicoctlv2 && \
+	  touch dist/calicoctlv2
+	-docker rm -f calicoctlv2
+
+## Run etcd as a container (calico-etcd)
+run-etcd: stop-etcd
+	docker run --detach \
+	--net=host \
+	--entrypoint=/usr/local/bin/etcd \
+	--name calico-etcd quay.io/coreos/etcd:v3.1.7 \
+	--advertise-client-urls "http://$(LOCAL_IP_ENV):2379,http://127.0.0.1:2379,http://$(LOCAL_IP_ENV):4001,http://127.0.0.1:4001" \
+	--listen-client-urls "http://0.0.0.0:2379,http://0.0.0.0:4001"
+
+
+.PHONY: stop-etcd
+stop-etcd:
+	@-docker rm -f calico-etcd
+
+# This depends on clean to ensure that dependent images get untagged and repulled
+.PHONY: semaphore
+semaphore: clean
+	# Clean up unwanted files to free disk space.
+	bash -c 'rm -rf /home/runner/{.npm,.phpbrew,.phpunit,.kerl,.kiex,.lein,.nvm,.npm,.phpbrew,.rbenv} /usr/local/golang /var/lib/mongodb'
+
+	# Run the containerized tests first.
+	$(MAKE) test-containerized st
+
+	$(MAKE) calico/upgrade
+
+	# Make sure that calico-upgrade builds cross-platform.
+	$(MAKE) dist/calico-upgrade-darwin-amd64 dist/calico-upgrade-windows-amd64.exe
+
+release: clean
+ifndef VERSION
+	$(error VERSION is undefined - run using make release VERSION=vX.Y.Z)
+endif
+	git tag $(VERSION)
```

Comment:
wouldn't the git tag for the calico directory take precedence? this looks correct if calico-upgrade had its own repo, but incorrect for a subdir off of calico.